[Tests] Failure to shutdown SGX-LKL after lthread_detach2()

There is an random enclave error during the shutdown sequence that makes tests fail:
```
Execution timeout: 60
Will run tests for run-hw
/opt/sgx-lkl-debug/bin/sgx-lkl-disk create --size=5M --docker=./Dockerfile sgx-lkl-rootfs.img
Using detected host DNS server in /etc/resolv.conf: 168.63.129.16
Creating sgx-lkl-rootfs.img from Dockerfile ./Dockerfile...
Building Docker image...
Creating disk image file from Docker container...
Succesfully created sgx-lkl-rootfs.img.
SGXLKL_VERBOSE=1 SGXLKL_KERNEL_VERBOSE=1 /opt/sgx-lkl-debug/bin/sgx-lkl-run-oe --hw-debug sgx-lkl-rootfs.img helloworld
[   SGX-LKL  ] SGX-LKL (OE) 0.2.0-dev (bc148c0) LKL 5.3.0 [DEBUG build (-O0)] [HARDWARE DEBUG]
[   SGX-LKL  ] nproc=4 ETHREADS=8 CMDLINE="mem=32M"
[   SGX-LKL  ] HW TLS support: conf->fsgsbase=1
[   SGX-LKL  ] Registering disk 0 (path='sgx-lkl-rootfs.img', mnt='/', [RW    ])
[   SGX-LKL  ] No tap device specified, networking will not be available.
[   SGX-LKL  ] get_signed_libsgxlkl_path... result=/opt/sgx-lkl-debug/bin/../lib/libsgxlkl.so.signed
[   SGX-LKL  ] oe_create_enclave... result=0 (OE_OK)
[   SGX-LKL  ] sgxlkl_enclave_init(ethread_id=0)
[[  SGX-LKL ]] sgxlkl_enclave_init(): enter
[[  SGX-LKL ]] sgxlkl_enclave_show_attribute(): enclave base=0x7fbd80000000 size=2.000 GB
[[  SGX-LKL ]] sgxlkl_enclave_show_attribute(): enclave heap base=0x7fbd80b55000 size=1024.00 M end=0x7fbdc0b55000
[[  SGX-LKL ]] sgxlkl_enclave_init(): calling _dlstart_c()
[[  SGX-LKL ]] _register_enclave_signal_handlers(): Registering OE exception handler...
[[  SGX-LKL ]] lkl_start_init(): kernel command line: 'mem=32M console=hvc0'
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() called
[    0.000000] Linux version 5.3.0+ (AzDevOps@cisca24ba0001LK) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #1 Sun Jun 28 03:04:33 UTC 2020
[    0.000000] memblock address range: 0x7fbdbe8d4000 - 0x7fbdc08d4000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 8080
[    0.000000] Kernel command line: mem=32M console=hvc0 virtio_mmio.device=256@0x1000000:1
[    0.000000] Dentry cache hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory available: 32176k/32768k RAM
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 4096
[    0.000000] lkl: irqs initialized
[    0.000000] clocksource: lkl: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000000] lkl: time and timers initialized (irq2)
[    0.000503] printk: console [hvc0] enabled
[    0.000503] pid_max: default: 4096 minimum: 301
[    0.000503] Mount-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.000503] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.002580] devtmpfs: initialized
[    0.006416] random: get_random_bytes called from 0x00007fbd8066ef6a with crng_init=0
[    0.006416] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.006416] futex hash table entries: 16 (order: -4, 384 bytes, linear)
[    0.006416] xor: automatically using best checksumming function   8regs     
[    0.006970] NET: Registered protocol family 16
[    0.010242] clocksource: Switched to clocksource lkl
[    0.010799] NET: Registered protocol family 2
[    0.011353] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.011353] TCP established hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.011353] TCP bind hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.011353] TCP: Hash tables configured (established 512 bind 512)
[    0.011353] UDP hash table entries: 128 (order: 0, 4096 bytes, linear)
[    0.011353] UDP-Lite hash table entries: 128 (order: 0, 4096 bytes, linear)
[    0.011353] NET: Registered protocol family 1
[    0.011353] virtio-mmio: Registering device virtio-mmio.0 at 0x1000000-0x10000ff, IRQ 1.
[    0.011909] workingset: timestamp_bits=62 max_order=14 bucket_order=0
[    0.012464] NET: Registered protocol family 38
[    0.012464] io scheduler mq-deadline registered
[    0.012464] io scheduler kyber registered
[    0.090329] software IO TLB: mapped [mem 0x7fbe76999000-0x7fbe7a999000] (64MB)
[    0.092549] loop: module loaded
[    0.092549] device-mapper: ioctl: 4.40.0-ioctl (2019-01-18) initialised: dm-devel@redhat.com
[    0.093106] wireguard: WireGuard 0.0.20191219 loaded. See www.wireguard.com for information.
[    0.093663] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.093663] NET: Registered protocol family 10
[    0.096345] Segment Routing with IPv6
[    0.096345] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.104614] This architecture does not have kernel memory protection.
[    0.104614] Run /init as init process
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() finished
[[  SGX-LKL ]] lkl_start_init(): creating LKL termination thread
[[  SGX-LKL ]] create_lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_start_init(): calling lkl_mount_virtial()
[[  SGX-LKL ]] lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_termination_thread(): Performed LKL syscall to get host task allocated (pid=36)
[[  SGX-LKL ]] lkl_start_init(): calling init_random()
[[  SGX-LKL ]] init_random(): Adding entropy to entropy pool
[    0.105728] random: crng init done
[[  SGX-LKL ]] init_enclave_clock(): Setting enclave realtime clock
[[  SGX-LKL ]] wg0 has public key b5JXGnyXeoxWCHR31F+bxqcSfx3x0e/64gTpSX+B0iw=
[    0.107399] virtio_blk virtio1: [vda] 10240 512-byte logical blocks (5.24 MB/5.00 MiB)
[[  SGX-LKL ]] lkl_mount_disk(): lkl_mount_disk(dev="/dev/vda", mnt="/mnt/vda", ro=0)
[    0.113444] EXT4-fs (vda): mounted filesystem without journal. Opts: 
[[  SGX-LKL ]] lkl_mount_disks(): Set working directory: /
[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: helloworld
Hello SGX-LKL-OE World!
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=0)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] display_mount_table(): ========= /proc/mounts ===========
/dev/vda / ext4 rw,relatime 0 0
devtmpfs /dev devtmpfs rw,relatime 0 0
tmpfs /dev/shm tmpfs rw,relatime 0 0
tmpfs /tmp tmpfs rw,relatime,mode=777 0 0
tmpfs /mnt tmpfs rw,relatime,mode=777 0 0
none /sys sysfs rw,relatime 0 0
tmpfs /run tmpfs rw,relatime,mode=700 0 0
proc /proc proc rw,relatime 0 0
[[  SGX-LKL ]] display_mount_table(): ==================================
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_umount_timeout("/", MNT_DETACH, 2000)
[[  SGX-LKL ]] display_mount_table(): /proc/mounts cannot be accessed
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[    0.119252] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
[   SGX-LKL  ] ethread (4: 0) ethread (5: 0) ethread (6: 0) 2020-06-28T03:38:37.000000Z [(H)ERROR] tid(0x7fbe727d2700) | :OE_ENCLAVE_ABORTING [/agent/_work/1/s/openenclave/host/calls.c:oe_call_enclave_function_by_table_id:91]
ethread (3: 19) [   SGX-LKL  ] FAIL: sgxlkl_ethread_init() failed (id=3 result=19 (OE_ENCLAVE_ABORTING))
ethread (7: 0) Makefile:29: recipe for target 'run-hw' failed
make: *** [run-hw] Error 1
make run-hw: FAILED WITH EXIT CODE: 2
Test run completed with EXIT CODE 2
```

The following code is executed before the init ethread leaves the enclave:
```
    /* Free the shutdown semaphore late in the shutdown sequence */
    sgxlkl_host_ops.sem_free(termination_sem);

    sgxlkl_free_config(sgxlkl_enclave);

    lthread_exit(NULL);
```

We need to figure out what could be causing the enclave abort. How can we output details why OE is aborting the enclave?

(cc: @mikbras @jhand2)

@prp this also failed for:
containers-cc-(debug)-(run-sw)-(8-ethreads)
```
Execution timeout: 300
Will run tests for run-sw
/opt/sgx-lkl-debug/bin/sgx-lkl-disk create --size=100M --docker=./Dockerfile rootfs.img
Using detected host DNS server in /etc/resolv.conf: 168.63.129.16
Creating rootfs.img from Dockerfile ./Dockerfile...
Building Docker image...
Creating disk image file from Docker container...
Succesfully created rootfs.img.
cp rootfs.img data.img
/opt/sgx-lkl-debug/bin/sgx-lkl-docker build --name=python-helloworld-cc --host-cfg=host-config.json --app-cfg=app-config.json
SGX-LKL root: /opt/sgx-lkl-debug/bin/..
Creating confidential container 'python-helloworld-cc'
Using temporary directory sgxlkl_docker_context_Yz45
Sending build context to Docker daemon  209.7MB
Step 1/14 : FROM ubuntu:18.04
 ---> 8e4ce0a6ce69
Step 2/14 : RUN apt-get update && apt-get install -y ca-certificates
 ---> Using cache
 ---> e1e2e084d5f9
Step 3/14 : RUN mkdir /prep &&     cd /prep &&     mkdir etc &&     mkdir -p home/sgxlkl &&     ls -al home/sgxlkl &&     mkdir tmp &&     chmod 777 tmp &&     mkdir -p etc/ssl/certs &&     cp /etc/ssl/certs/ca-certificates.crt etc/ssl/certs &&     ls -al etc/ssl/certs
 ---> Using cache
 ---> 272d6f5363cc
Step 4/14 : COPY . /prep/home/sgxlkl
 ---> e6bae37b15bc
Step 5/14 : RUN chmod -R 777 /prep/home/sgxlkl
 ---> Running in 14bf311ceea1
Removing intermediate container 14bf311ceea1
 ---> 4419bca51136
Step 6/14 : FROM scratch
 ---> 
Step 7/14 : ENV HOME=/home/sgxlkl
 ---> Using cache
 ---> a06f906bd194
Step 8/14 : ENV TMPDIR=/tmp
 ---> Using cache
 ---> f8a858343138
Step 9/14 : ENV AZDCAP_CACHE=${TMPDIR}
 ---> Using cache
 ---> 04198d2fc690
Step 10/14 : COPY --from=0 /prep /
 ---> 2f328a95c089
Step 11/14 : USER 1000:1000
 ---> Running in 40702c0c6f9a
Removing intermediate container 40702c0c6f9a
 ---> 72a3a4a6783b
Step 12/14 : WORKDIR ${HOME}
 ---> Running in e54db8db0587
Removing intermediate container e54db8db0587
 ---> bb90ecf24d42
Step 13/14 : ENTRYPOINT ["/opt/sgx-lkl/lib/external/ld-linux-x86-64", "/opt/sgx-lkl/bin/sgx-lkl-run-oe",    "--enclave-image=/opt/sgx-lkl/lib/libsgxlkl.so.signed",    "--host-config=host-cfg.json",    "--app-config=app-cfg.json"]
 ---> Running in 7f21d56865ea
Removing intermediate container 7f21d56865ea
 ---> c2ea0b05f581
Step 14/14 : CMD ["--hw-release"]
 ---> Running in d9281df763d3
Removing intermediate container d9281df763d3
 ---> 10e49cb94d68
Successfully built 10e49cb94d68
Successfully tagged python-helloworld-cc:latest
Done. Tagged 'python-helloworld-cc'.
touch python-helloworld-cc.stamp
# TODO -t as work-around for https://github.com/lsds/sgx-lkl/issues/158
docker run -t --rm --user 1001 --device /dev/net/tun --network host -v /opt/sgx-lkl-debug:/opt/sgx-lkl -e SGXLKL_VERBOSE=1 -e SGXLKL_KERNEL_VERBOSE=1 -e SGXLKL_TAP=sgxlkl_tap0 -e AZDCAP_DEBUG_LOG_LEVEL=INFO python-helloworld-cc --sw-debug
[   SGX-LKL  ] SGX-LKL (OE) 0.2.0-dev (8c223be) LKL 5.3.0 [DEBUG build (-O0)] [SOFTWARE DEBUG]
[   SGX-LKL  ] nproc=8 ETHREADS=8 CMDLINE="mem=32M"
[   SGX-LKL  ] HW TLS support: conf->fsgsbase=0
[   SGX-LKL  ] Registering disk 0 (path='disk_0.img', mnt='/', [RW    ])
[   SGX-LKL  ] Registering disk 1 (path='disk_1.img', mnt='/data', [RW    ])
[   SGX-LKL  ] get_signed_libsgxlkl_path... result=/opt/sgx-lkl/lib/libsgxlkl.so.signed
[   SGX-LKL  ] oe_create_enclave... result=0 (OE_OK)
[   SGX-LKL  ] sgxlkl_enclave_init(ethread_id=0)
[[  SGX-LKL ]] sgxlkl_enclave_init(): enter
[[  SGX-LKL ]] _sgxlkl_enclave_show_attribute(): enclave base=0x7fe400000000 size=2.0 GiB
[[  SGX-LKL ]] _sgxlkl_enclave_show_attribute(): enclave heap base=0x7fe400b41000 size=1024.0 MiB end=0x7fe440b41000
[[  SGX-LKL ]] sgxlkl_enclave_init(): calling _dlstart_c()
[[  SGX-LKL ]] _register_enclave_signal_handlers(): Registering OE exception handler...
[[  SGX-LKL ]] __libc_init_enclave(): calling enclave_mman_init()
[[  SGX-LKL ]] __libc_init_enclave(): calling _lthread_sched_init()
[[  SGX-LKL ]] startmain(): enter
[[  SGX-LKL ]] lkl_start_init(): calling register_lkl_syscall_overrides()
[[  SGX-LKL ]] lkl_start_init(): calling initialize_enclave_event_channel()
[[  SGX-LKL ]] lkl_start_init(): calling lkl_virtio_console_add()
[[  SGX-LKL ]] lkl_start_init(): kernel command line: 'mem=32M console=hvc0'
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() called
[    0.000000] Linux version 5.3.0+ (AzDevOps@cisca24ba0001UH) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #1 Sat Jul 4 01:04:46 UTC 2020
[    0.000000] memblock address range: 0x7fe43e840000 - 0x7fe440840000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 8080
[    0.000000] Kernel command line: mem=32M console=hvc0 virtio_mmio.device=256@0x1000000:1 virtio_mmio.device=273@0x2000000:2
[    0.000000] Dentry cache hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 2048 (order: 2, 16384 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory available: 32176k/32768k RAM
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 4096
[    0.000000] lkl: irqs initialized
[    0.000000] clocksource: lkl: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000000] lkl: time and timers initialized (irq3)
[    0.005592] printk: console [hvc0] enabled
[    0.005592] pid_max: default: 4096 minimum: 301
[    0.006168] Mount-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.006168] Mountpoint-cache hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.007275] devtmpfs: initialized
[    0.015067] random: get_random_bytes called from 0x00007fe40065f7da with crng_init=0
[    0.015067] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.015592] futex hash table entries: 16 (order: -4, 384 bytes, linear)
[    0.015592] xor: automatically using best checksumming function   8regs     
[    0.015592] NET: Registered protocol family 16
[    0.020043] clocksource: Switched to clocksource lkl
[    0.020604] NET: Registered protocol family 2
[    0.021165] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.021739] TCP established hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.021739] TCP bind hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.022300] TCP: Hash tables configured (established 512 bind 512)
[    0.022300] UDP hash table entries: 128 (order: 0, 4096 bytes, linear)
[    0.022862] UDP-Lite hash table entries: 128 (order: 0, 4096 bytes, linear)
[    0.022862] NET: Registered protocol family 1
[    0.023425] virtio-mmio: Registering device virtio-mmio.0 at 0x1000000-0x10000ff, IRQ 1.
[    0.023985] virtio-mmio: Registering device virtio-mmio.1 at 0x2000000-0x2000110, IRQ 2.
[    0.023985] workingset: timestamp_bits=62 max_order=14 bucket_order=0
[    0.026789] NET: Registered protocol family 38
[    0.026789] io scheduler mq-deadline registered
[    0.027342] io scheduler kyber registered
[    0.043611] loop: module loaded
[    0.044178] device-mapper: ioctl: 4.40.0-ioctl (2019-01-18) initialised: dm-devel@redhat.com
[    0.044742] wireguard: WireGuard 0.0.20191219 loaded. See www.wireguard.com for information.
[    0.044742] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.045306] NET: Registered protocol family 10
[    0.045869] Segment Routing with IPv6
[    0.045869] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.047012] This architecture does not have kernel memory protection.
[    0.047012] Run /init as init process
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() finished
[[  SGX-LKL ]] lkl_start_init(): creating LKL termination thread
[[  SGX-LKL ]] create_lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_start_init(): calling lkl_mount_virtial()
[[  SGX-LKL ]] lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_start_init(): calling init_random()
[[  SGX-LKL ]] init_random(): Adding entropy to entropy pool
[[  SGX-LKL ]] lkl_termination_thread(): Performed LKL syscall to get host task allocated (pid=36)
[    0.048196] random: crng init done
[[  SGX-LKL ]] init_enclave_clock(): Setting enclave realtime clock
[    0.190399] software IO TLB: mapped [mem 0x7fe4dc43a000-0x7fe4e043a000] (64MB)
[[  SGX-LKL ]] wg0 has public key tcb3ZEp67m4nczH0LHBd3RoRnj99kc5d+gKPnjnpOAk=
[    0.192650] virtio_blk virtio2: [vda] 204800 512-byte logical blocks (105 MB/100 MiB)
[[  SGX-LKL ]] lkl_mount_disk(): lkl_mount_disk(dev="/dev/vda", mnt="/mnt/vda", ro=0)
[    0.194340] EXT4-fs (vda): mounted filesystem without journal. Opts: 
[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: /usr/bin/python3
Hello world!
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/urllib3/response.py", line 397, in _error_catcher
    yield
  File "/usr/lib/python3.7/site-packages/urllib3/response.py", line 479, in read
    data = self._fp.read(amt)
  File "/usr/lib/python3.7/http/client.py", line 457, in read
    n = self.readinto(b)
  File "/usr/lib/python3.7/http/client.py", line 501, in readinto
    n = self.fp.readinto(b)
  File "/usr/lib/python3.7/socket.py", line 589, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib/python3.7/ssl.py", line 1071, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib/python3.7/ssl.py", line 929, in read
    return self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/requests/models.py", line 750, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/usr/lib/python3.7/site-packages/urllib3/response.py", line 531, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/lib/python3.7/site-packages/urllib3/response.py", line 496, in read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
  File "/usr/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/lib/python3.7/site-packages/urllib3/response.py", line 402, in _error_catcher
    raise ReadTimeoutError(self._pool, None, 'Read timed out.')
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='www.microsoft.com', port=443): Read timed out.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/python-helloworld.py", line 5, in <module>
    response = requests.get("https://www.microsoft.com/", timeout=5)
  File "/usr/lib/python3.7/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/usr/lib/python3.7/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 668, in send
    history = [resp for resp in gen] if allow_redirects else []
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 668, in <listcomp>
    history = [resp for resp in gen] if allow_redirects else []
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 247, in resolve_redirects
    **adapter_kwargs
  File "/usr/lib/python3.7/site-packages/requests/sessions.py", line 686, in send
    r.content
  File "/usr/lib/python3.7/site-packages/requests/models.py", line 828, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
  File "/usr/lib/python3.7/site-packages/requests/models.py", line 757, in generate
    raise ConnectionError(e)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='www.microsoft.com', port=443): Read timed out.
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=1)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] display_mount_table(): ========= /proc/mounts ===========
/dev/vda / ext4 rw,relatime 0 0
devtmpfs /dev devtmpfs rw,relatime 0 0
tmpfs /dev/shm tmpfs rw,relatime 0 0
tmpfs /tmp tmpfs rw,relatime,mode=777 0 0
tmpfs /mnt tmpfs rw,relatime,mode=777 0 0
none /sys sysfs rw,relatime 0 0
tmpfs /run tmpfs rw,relatime,mode=700 0 0
proc /proc proc rw,relatime 0 0
[[  SGX-LKL ]] display_mount_table(): ==================================
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_umount_timeout("/", MNT_DETACH, 2000)
[[  SGX-LKL ]] display_mount_table(): /proc/mounts cannot be accessed
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[    5.731407] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
[   SGX-LKL  ] init (0: 0 exit=1) ethread (4: 0) ethread (6: 0) ethread (3: 0) ethread (5: 0) ethread (7: 0) ethread (2: 0) ethread (1: 0) 
[   SGX-LKL  ] oe_terminate_enclave... done
[   SGX-LKL  ] SGX-LKL-OE exit: exit_status=1
Makefile:52: recipe for target 'run-sw' failed
make: *** [run-sw] Error 1
make run-sw: FAILED WITH EXIT CODE: 2
Test run completed with EXIT CODE 2
```

--- a/src/enclave/enclave_signal.c
+++ b/src/enclave/enclave_signal.c
@@ -8,6 +8,7 @@
 #include <asm-generic/ucontext.h>
 
 #include <lkl_host.h>
+#include <lkl/setup.h>
 #include <string.h>
 #include "enclave/sgxlkl_t.h"
 #include "enclave/lthread.h"
@@ -151,10 +152,11 @@ static uint64_t sgxlkl_enclave_signal_handler(
 #endif
 
         /**
-         * If LKL has not yet been initialised, we cannot handle the
-         * exception and fail instead.
+         * If LKL has not yet been initialised or is terminating (and thus no
+         * longer accepts signals), we cannot handle the exception and fail
+         * instead.
          */
-        if (!lkl_is_running())
+        if (!lkl_is_running() || is_lkl_terminating())
         {
 #ifdef DEBUG
             sgxlkl_error("Exception received before LKL can handle it. "

--- a/src/enclave/enclave_signal.c
+++ b/src/enclave/enclave_signal.c
@@ -159,7 +159,7 @@ static uint64_t sgxlkl_enclave_signal_handler(
         if (!lkl_is_running() || is_lkl_terminating())
         {
 #ifdef DEBUG
-            sgxlkl_error("Exception received before LKL can handle it. "
+            sgxlkl_error("Exception received but LKL is unable to handle it. "
                          "Printing stack trace saved by exception handler:\n");
             /**
              * Since we cannot unwind the frames in the OE exception handler,
@@ -171,7 +171,8 @@ static uint64_t sgxlkl_enclave_signal_handler(
 
             struct lthread* lt = lthread_self();
             sgxlkl_fail(
-                "Exception %s received before LKL is running (lt->tid=%i [%s] "
+                "Exception %s received before LKL initialisation/after LKL "
+                "shutdown (lt->tid=%i [%s] "
                 "code=%i "
                 "addr=0x%lx opcode=0x%x "
                 "ret=%i)\n",

--- a/src/include/lkl/setup.h
+++ b/src/include/lkl/setup.h
@@ -17,4 +17,7 @@ void lkl_mount_disks(
 /* Shutdown the running LKL kernel */
 void lkl_terminate(int exit_status);
 
+/* Return if LKL is currently terminating */
+bool is_lkl_terminating();
+
 #endif /* SETUP_H */

--- a/src/lkl/setup.c
+++ b/src/lkl/setup.c
@@ -1182,7 +1182,7 @@ static void display_mount_table()
 static struct lkl_sem* termination_sem;
 
 /* Record whether we are terminmating LKL */
-static _Atomic(bool) is_lkl_terminating = false;
+static _Atomic(bool) _is_lkl_terminating = false;
 
 /* Function to carry out the shutdown sequence */
 static void* lkl_termination_thread(void* args)
@@ -1344,14 +1344,20 @@ void lkl_terminate(int exit_status)
      * We only want to trigger the shutdown once. Since we are shutting down
      * the applicaton and the kernel, many other threads will be exiting now.
      */
-    if (!is_lkl_terminating)
+    if (!_is_lkl_terminating)
     {
         SGXLKL_VERBOSE("terminating LKL (exit_status=%i)\n", exit_status);
-        is_lkl_terminating = true;
+        _is_lkl_terminating = true;
         sgxlkl_exit_status = exit_status;
         /* Wake up LKL termination thread to carry out the work. */
         sgxlkl_host_ops.sem_up(termination_sem);
     }
+}
+
+/* Return if LKL is currently terminating */
+bool is_lkl_terminating()
+{
+    return _is_lkl_terminating;
 }
 
 static void init_enclave_clock()

--- a/src/sched/lthread.c
+++ b/src/sched/lthread.c
@@ -263,7 +263,7 @@ void lthread_run(void)
         if (_lthread_should_stop)
         {
             SGXLKL_TRACE_THREAD(
-                "[tid=%-3d] lthread_run(): quiting\n", lt ? lt->tid : -1);
+                "[tid=%-3d] lthread_run(): quitting\n", lt ? lt->tid : -1);
             break;
         }
     }

--- a/src/sched/lthread.c
+++ b/src/sched/lthread.c
@@ -225,9 +225,8 @@ void lthread_run(void)
                 pauses = sleepspins;
                 a_dec(&schedqueuelen);
                 SGXLKL_TRACE_THREAD(
-                    "[tid=%-3d] lthread_run() lthread_resume (dequeue sched "
-                    "queue) \n",
-                    lt->tid);
+                    "[tid=%-3d] lthread_run(): lthread_resume (dequeue)\n",
+                    lt ? lt->tid : -1);
                 _lthread_resume(lt);
             }
 
@@ -263,7 +262,8 @@ void lthread_run(void)
         /* Break out of scheduler loop when enclave is terminating */
         if (_lthread_should_stop)
         {
-            SGXLKL_TRACE_THREAD("[tid=%-3d] lthread_run() quiting.\n", lt->tid);
+            SGXLKL_TRACE_THREAD(
+                "[tid=%-3d] lthread_run(): quiting\n", lt ? lt->tid : -1);
             break;
         }
     }

